Add `lock` and `unlock` commands for issues and pull requests

chemotaxis opened the pull request:
Fixes #5020.

This pull request adds `lock` and `unlock` to issues and pull requests.

`unlock` will unlock a conversation if it was previously locked.  Otherwise, it will do nothing.

`lock` will lock a conversation if it was previously unlocked.  You can optionally specify a reason for locking from a set of fixed reasons.

If the issue or pull request is already locked, `gh` will prompt you to confirm if you want to "relock" a conversation or abort the "relock" if you forgot or didn't know the conversation was already locked.  The primary use for relocking will probably be to change the lock reason.

I tried just sending a lock request again with an updated reason, but nothing changes.  You first have to unlock the conversation, then lock it with the updated reason.

The error messages are parameterized on several variables including the parent command (`issue` or `pr`), what the intended action was (lock or unlock), the reason for locking, etc.  For example, if __`issue #1`__ exists, but we used the command `gh pr lock 1`, the error message will say that __`issue #1`__ exists, but not __`pull request #1`__.  It then suggests to use `gh issue lock 1` instead.
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

chemotaxis committed:
Stub out files for lock command

Lock will lock and unlock both issues and pull requests

chemotaxis committed:
Draft first design

As originally designed in the issue discussion, a single function
`NewCmdLock()` with a parameter to lock or unlock was proposed.
However, after playing around with a couple different designs, it seems
best to create two separate public functions and one private function to
do the common work.

Using two public functions seems to make sense because the api for
locking is different from the api for unlocking.  Therefore, the
documentation for both are different and keeping them in separate
functions would make it easier to maintain the documentation.

chemotaxis committed:
Added lock subcommands to pr command

chemotaxis committed:
Other changes

- Changed function to method
- Moved additional common options to method
- Remove redundant documentation
    - Cobra sets documentation in the Command struct.

chemotaxis committed:
Export return values for Issue.Typename

These are needed to know if an issue is an actual issue or a pull
request.

chemotaxis committed:
Map parent commands to typenames

chemotaxis committed:
Add lock-related values to issue

Add queries for "ActiveLockReason" and "Locked"

chemotaxis committed:
Rearrange

chemotaxis committed:
Get graphql queries and mutations working

- Fix error if found an issue while using `gh pr lock/unlock` or vice versa
- Added additional types
- Used githubv4 types
- Added "relock" state
    - If the conversation is already locked you have two choices: try to
      lock it again or do nothing.  Do nothing is easy.  But, if you
      want to change the lock reason, you need to first unlock the
      conversation and then lock it again.
- Added survey to confirm if you want to relock
- Added formatted print statements

chemotaxis committed:
Modify documentation if called from issue or pr

chemotaxis committed:
Skip unlock api call if already unlocked

chemotaxis committed:
Remove opts.PadlockState

I noticed that PadlockState didn't really have anything to do with the
LockOptions and it was easy to call an incorrect locking function that
didn't match the PadlockState.

Now, you pass in the state as an argument and you simply call the
appropriate function instead of setting PadlockState and then calling
the correct function.

- Other touch ups and refactoring

chemotaxis committed:
Use `if/else if` on error checking

Use switch statements for regular branching code.

chemotaxis committed:
Polished error message

chemotaxis committed:
Restructured switch statement

As much as I like keeping statements as flat as possible, this
not-as-flat revision just seems easier to read.

chemotaxis committed:
Polished things

chemotaxis committed:
Refactor things

- Switch to underscores
- Revise error message

chemotaxis committed:
Remove opts.PadlockState

Rather than saving the intended lock state and calling a method
depending on the lock state, just call the method directly.  By the time
you need to the padlock state, you already know which method to use; no
need to first change the lock state than call the method.

Also, refactored print/error messages that are conditional.

chemotaxis committed:
Add failure icon

chemotaxis commented:
I apologize if the history on this pull request is a little messy, but I went through several drafts before arriving at this current version.  I haven't written any tests yet, but I wanted to see if the overall behavior was acceptable.

Some notes:

There seem to be several methods with overlapping functionality in `issue/shared` and `pr/shared`, so it was a little confusing at first to find the most appropriate function to request information about an issue or pull request.

It was also somewhat tricky to address two separate functions (lock/unlock) and two separate types (issues/pull requests) simultaneously.  Hopefully my current solution addresses everything fairly cleanly.

Finally, since the current version of `gh` doesn't support querying/mutating the lock state or active lock reason, I needed to add a couple new fields to the `Issue` type.  Consequently, `gh` doesn't display the locked state or lock reason.  I plan to create a new issue in order to add lock-related information to `issue view`, `pr view`, at least.

chemotaxis committed:
Merge remote-tracking branch 'upstream/trunk' into gh-issue-lock

Update a somewhat stale feature branch with the latest stable commits.

vilmibm commented on pkg/cmd/issue/lock/lock.go:
```diff
@@ -0,0 +1,313 @@
+		},
+	}
+
+	gql := graphql.NewClient(ghinstance.GraphQLEndpoint(repo.RepoHost()), httpClient)
```
Should use `api.NewClientFromHTTP(httpClient)`

vilmibm commented on pkg/cmd/issue/lock/lock.go:
```diff
@@ -0,0 +1,313 @@
+		},
+	}
+
+	gql := graphql.NewClient(ghinstance.GraphQLEndpoint(repo.RepoHost()), httpClient)
```
Should use `api.NewClientFromHTTP(httpClient)`

vilmibm commented on pkg/cmd/issue/lock/lock.go:
```diff
@@ -0,0 +1,313 @@
+		Default: true,
+	}
+
+	err := survey.AskOne(shouldRelock, &relocked)
```
This should:

- use the new `Prompter`
- not prompt if STDIN is not a tty (see `IOstreams.CanPrompt` )

vilmibm commented on pkg/cmd/issue/lock/lock_test.go:
```diff
@@ -0,0 +1 @@
+package lock
```
needs tests

vilmibm reviewed (changes requested):
This is a great start. I'm cool with the UX and appreciate the effort put into making this work across issues and prs.

chemotaxis committed:
Merge remote-tracking branch 'upstream/trunk' into gh-issue-lock

Add recent changes, especially changes to `api`.

chemotaxis committed:
Remove direct calls to graphql library

Similar to commit 45f1a71c8bae265ecd9985812e32a6f9d34027c5

chemotaxis committed:
Update comments

chemotaxis committed:
Alphabetize fields

chemotaxis committed:
Implicitly return the nil value when no lock reason given

While I usually like explicitly setting all values, getting rid of
setting the empty string streamlines map construction and testing.

The reasonsMap will return the nil value because the empty string is not
in the map.

chemotaxis committed:
Test reasonsMap construction

Make sure reasons were added in the correct order.

Upgradeddevil05 reviewed (commented):
[https://pay.google.com/gp/w/u/0/home/activity?sctid=1478388430540775](https://pay.google.com/gp/w/u/0/home/activity?sctid=1478388430540775)

chemotaxis committed:
Merge remote-tracking branch 'upstream/trunk' into gh-issue-lock

Resolved conflicts mainly around pinning comments.

Commit:
Merge remote-tracking branch 'origin/trunk' into gh-issue-lock

Commit:
wip: tweak output, update to Prompter, start on tests

Commit:
WIP on adding tests, add prompt, tweak things

Commit:
further WIP

Commit:
WIP on tests

Commit:
WIP tests

Commit:
todos

Commit:
stub out rest of tests

Commit:
WIP all but relocking

Commit:
finish tests

vilmibm commented:
I've:

- added an interactive picker for reason if TTY and none selected
- simplified error messages when stdout is not TTY
- tweaked some wording
- added a full test suite
